core/src/main/java/io/ultrabrew/metrics/data/ConcurrentMonoidLongTable.java

```diff
@@ -294,11 +296,30 @@ long index(final String[] tags, final boolean isReading) {
                 // grow tag set
                 synchronized (this) {
                   if (tagIndex >= tagSets.length) {
+                    boolean isEqual = false;
+                    if(tagIndex == tagSets.length) {
+                      isEqual = true;
+                    }
                     final int oldLength = tagSets.length;
                     final int newLength =
                         oldLength > TAGSETS_MAX_INCREMENT ? oldLength + TAGSETS_MAX_INCREMENT
                             : oldLength * 2;
-                    tagSets = Arrays.copyOf(tagSets, newLength);
+                    String[][] newTagSet = new String[newLength][];
+                    int cnt = 0;
+                    for (int i = 0; i < tagSets.length; i++) {
+                      // Skip the entries that would have been freed up (stale)
+                      // System.arraycopy would be faster, but can't use it here due to 
+                      // the conditional inclusion 
+                      if ((tagSets[i].length > 0 && tagSets[i][0] != null)
+                          || (tagSets[i].length == 0)) {
+                        String[] newTag = tagSets[i];
+                        newTagSet[cnt++] = newTag;
+                      }
+                    }
+                    tagSets = newTagSet;
+                    // reset the usedOffset length;
+                    unsafe.putIntVolatile(this, usedOffset, isEqual ? cnt + 1 : cnt + 2);
```

Comment:
There's a race here between `tagSets` being replaced and `usedOffset` getting rewritten - that seems like a problem unless something elsewhere makes this safe?

Reply:
This is wrapped in a synchronized block, so we should be ok.

Reply:
Despite the synchronized block there's a race where one thread can get wrong value for `tagIndex` when another thread compacts the array. This used to rely heavily on the fact that `usedOffset` could never decrease, making it safe for that requires some thought.

Reply:
We would keep incrementing the `usedOffset` and enter this block only when `usedOffset` value is more than `tagSets` length. One thread would get a chance to expand the array and the other thread waiting on the lock will check for the array length again and since the array is expanded, it will break out of the condition. Given we do not shrink the array, do you think of a case the race would still happen?

Reply:
The DCLP is fine here and protects the assignment. But now won't the array just keep growing forever? Albeit with some empty slots? We also need a way to actually compact the array and relieve pressure.

The bigger issue GSK and I talked about is that the nulling of the tag set is done by the reporter and the tag set is shared amongst reporters so one could null it while the other doesn't and that would goof things up. So there is a bit of work to be redone here.

Reply:
The `tagSets` array will be compacted by excluding the elements that are marked for deletion.
https://github.com/ultrabrew/metrics/pull/28/files#diff-fdded27fe21d54ab627b000d6b884cc8R320
and https://github.com/ultrabrew/metrics/pull/28/files#diff-fdded27fe21d54ab627b000d6b884cc8R320

Reply:
Consider this interleaving of events:

Preconditions:
 `tagSets.length == 2`
 `usedOffset == 2`
 `tagSets` contains 1 stale entry

Thread 1:

```java
int tagIndex = unsafe.getAndAddInt(this, usedOffset, 1);
// tagIndex == 2
```

Thread 2:

```java
int tagIndex = unsafe.getAndAddInt(this, usedOffset, 1);
// tagIndex == 3 
if (tagIndex >= tagSets.length) { 
  synchronized (this) { 
    if (tagIndex >= tagSets.length) { 
                  
      boolean isEqual = tagIndex == tagSets.length;
      final int oldLength = tagSets.length;
      final int newLength = 
          oldLength > TAGSETS_MAX_INCREMENT ? oldLength + TAGSETS_MAX_INCREMENT
              : oldLength * 2;
      String[][] newTagSet = new String[newLength][];
      int cnt = 0;
      for (int i = 0; i < tagSets.length; i++) { 
        String[] tagSeti = tagSets[i];
        if ((tagSeti.length > 0 && tagSeti[0] != null)
            || (tagSeti.length == 0)) { 
          String[] newTag = tagSeti;
          newTagSet[cnt++] = newTag;
        } 
      } 
      // cnt == 1, isEqual == false
      unsafe.putIntVolatile(this, usedOffset, isEqual ? cnt + 1 : cnt + 2);
      // usedOffset == 2 
      tagSets = newTagSet;
      tagIndex = isEqual ? cnt : cnt + 1;
      // tagIndex == 2 
    } 
  } 
} 
```


Thread 1:

```java
if (tagIndex >= tagSets.length) { 
  // skipped, condition is false
} 
```

Now both threads have `tagIndex == 2` and are about to write to `tagSets[tagIndex]`.

Reply:
@lonemeow yes, there's a race as you mentioned. Since the array would potentially be shrinked, there's not a clean way to handle this scenario, will close this pr out for now and will open it up once I have a solution.